llvm/lib/Analysis/ValueTracking.cpp

```diff
@@ -8890,6 +8870,62 @@ static Value *lookThroughCast(CmpInst *CmpI, Value *V1, Value *V2,
   return CastedTo;
 }
 
+/// Helps to match a select pattern in case of a type mismatch.
+///
+/// The function processes the case when type of true and false values of a
+/// select instruction differs from type of the cmp instruction operands because
+/// of a cast instruction. The function checks if it is legal to move the cast
+/// operation after "select". If yes, it returns the new second value of
+/// "select" (with the assumption that cast is moved):
+/// 1. As operand of cast instruction when both values of "select" are same cast
+/// instructions.
+/// 2. As restored constant (by applying reverse cast operation) when the first
+/// value of the "select" is a cast operation and the second value is a
+/// constant. It is implemented in lookThroughCastConst().
+/// 3. As one operand is cast instruction and the other is not. The operands in
+/// sel(cmp) are in different type integer.
+/// NOTE: We return only the new second value because the first value could be
+/// accessed as operand of cast instruction.
+static Value *lookThroughCast(CmpInst *CmpI, Value *V1, Value *V2,
+                              Instruction::CastOps *CastOp) {
+  auto *Cast1 = dyn_cast<CastInst>(V1);
+  if (!Cast1)
+    return nullptr;
+
+  *CastOp = Cast1->getOpcode();
+  Type *SrcTy = Cast1->getSrcTy();
+  if (auto *Cast2 = dyn_cast<CastInst>(V2)) {
+    // If V1 and V2 are both the same cast from the same type, look through V1.
+    if (*CastOp == Cast2->getOpcode() && SrcTy == Cast2->getSrcTy())
+      return Cast2->getOperand(0);
+    return nullptr;
+  }
+
+  auto *C = dyn_cast<Constant>(V2);
+  if (C)
+    return lookThroughCastConst(CmpI, V1, V2, CastOp);
```

Comment:
Please adjust `lookThroughCastConst` to avoid redundant checks.
```
static Value *lookThroughCastConst(CmpInst *CmpI, CastInst *Cast1, Constant *C,
                                    Instruction::CastOps CastOp)
```

Reply:
Do you mean remove
+  if (!Cast1)
+    return nullptr; 
and 
+  if (!C)
+    return nullptr;

I remained it just only for robustness of lookThroughCastConst, if some one will call lookThroughCastConst independently.

Reply:
@nikic @dtcxzyw Any other comments?

Reply:
@tianleliu The checks are not necessary if you pass through the correct types (the dyn_cast results).